src/us/kbase/workspace/database/mongo/MongoWorkspaceDB.java

```diff
@@ -2308,80 +2316,139 @@ private void saveData(
 	@Override
 	public void addDataToObjects(
 			final Collection<WorkspaceObjectData.Builder> objects,
-			final ByteArrayFileCacheManager dataManager)
+			final ByteArrayFileCacheManager dataManager,
+			final int backendScaling)
 			throws WorkspaceCommunicationException, TypedObjectExtractionException,
-				NoObjectDataException {
+				NoObjectDataException, InterruptedException {
 		noNulls(requireNonNull(objects, "objects"), "null found in objects");
+		requireNonNull(dataManager, "dataManager");
+		if (backendScaling < 1) {
+			throw new IllegalArgumentException("backendScaling must be > 0");
+		}
+		if (objects.isEmpty()) {
+			return;
+		}
 		final Map<String, List<WorkspaceObjectData.Builder>> chksum2obj = new HashMap<>();
 		for (final WorkspaceObjectData.Builder b: objects) {
 			if (!chksum2obj.containsKey(b.getObjectInfo().getCheckSum())) {
 				chksum2obj.put(b.getObjectInfo().getCheckSum(), new LinkedList<>());
 			}
 			chksum2obj.get(b.getObjectInfo().getCheckSum()).add(b);
 		}
-		final List<ByteArrayFileCache> toDestroy = new LinkedList<>();
+		final List<BackendFileCallable> callables = chksum2obj.entrySet().stream()
+				.map(e -> new BackendFileCallable(blob, e.getKey(), dataManager, e.getValue()))
+				.collect(Collectors.toList());
+		final ExecutorService eserv = Executors.newFixedThreadPool(
+				Math.min(backendScaling, chksum2obj.size()));
 		try {
-			for (final String chksum: chksum2obj.keySet()) {
-				// TODO PRL parallelize this code, make sure to clean up all threads' data 
-				// if one fails
-				final ByteArrayFileCache data;
-				try {
-					data = blob.getBlob(new MD5(chksum), dataManager);
-					toDestroy.add(data);
-				} catch (IOException e) {
-					throw new WorkspaceCommunicationException(e.getLocalizedMessage(), e);
-				} catch (BlobStoreCommunicationException e) {
-					throw new WorkspaceCommunicationException(e.getLocalizedMessage(), e);
-				} catch (BlobStoreAuthorizationException e) {
-					throw new WorkspaceCommunicationException(
-							"Authorization error communicating with the backend storage system",
-							e);
-				} catch (NoSuchBlobException e) {
-					final ObjectInformation info = chksum2obj.get(chksum).get(0).getObjectInfo();
-					throw new NoObjectDataException(String.format(
-							"No data present for object %s/%s/%s",
-							info.getWorkspaceId(), info.getObjectId(), info.getVersion()), e);
-				}
-				for (final WorkspaceObjectData.Builder b: chksum2obj.get(chksum)) {
-					/* might be subsetting the same object the same way multiple times, but
-					 * probably unlikely. If it becomes a problem memoize the subset
-					 */
-					b.withData(getDataSubSet(data, b.getSubsetSelection(), dataManager));
-				}
+			final List<Future<Void>> futures = eserv.invokeAll(
+					callables, OBJECT_DATA_FETCH_TIMEOUT_SEC, TimeUnit.SECONDS);
```

Comment:
An alternative implementation could be to have a global executor service and submit `backendScaling` meta jobs that all share a synchronized list of jobs and pull jobs off the queue in a loop. That would improve over the current implementation in two ways:

* The single executor service would allow capping the total number of threads talking to the backend per service and could be configured (probably in `deploy.cfg`, switching executors live seems iffy). Currently the effective number of threads is the server thread count (which is 500 for tomcat IIRC) * `backendScaling`.
* The meta jobs could cancel early if one job fails, and could even set a semaphore so all the meta jobs could fail. Currently all jobs run to completions regardless of the status of the other jobs.

All that being said, the current implementation is much simpler and probably perfectly fine, so YAGNI.